cilium/CFP-22215-mutual-auth-for-service-mesh.md

```diff
@@ -0,0 +1,466 @@
+* If there are two conflicting policies, one that requires authentication and one that does not require authentication, then authentication will be required
+* If there are two policies with conflicting authentication fields, reject the rule
+  * Log an error in the Cilium agent logs
+  * Emit a warning status to in-pod `cilium status` output
+  * Increment a metric, with tags
+* Gotchas / corner cases:
+  * Avoid requiring auth to kube-apiserver
+  * Avoid requiring auth to SPIRE
+  * Many more, still TBD, [#24552](https://github.com/cilium/cilium/issues/24552) is logged to investigate, and this document will be updated further once that work is done.
+
+
+### mTLS, SPIFFE, and SPIRE
+
+When building an mTLS solution, SPIFFE, the API for having workloads request an attested identity with a clear chain of custody, and SPIRE, the implementation of that API, are the state of the art. The SPIFFE project has spent a lot of time thinking deeply about how you can avoid common identity-sharing pitfalls like spoofing identities, getting identity details (commonly but not always X.509 TLS keypairs, but not always) to the workload security, and being able to easily rotate and revoke these identities when required.
+
+This proposal recommends that, rather than trying to build something that will take a while to handle all of the above sharp edges, we use SPIFFE and SPIRE as essential parts of the Cilium Service Mesh story, and associate the Cilium Security Identity on a 1:1 basis with a SPIFFE identity and the associated X.509 keypairs.
+
+Pros:
+* We’re using another CNCF project, which means that we’re both being good citizens, and avoiding building something ourselves.
+* We can avoid having to run down all the ways a fast, proof-of-concept implementation would create security issues in the future. There’s no such thing as temporary when it comes to APIs.
+* We effectively add a second factor to the Cilium Security Identity - the SPIFFE identity and the keypair in the SVID.
+* The certificates aren’t actually stored in the Kubernetes cluster as Secrets, so getting access to them is more difficult. For example, someone with filesystem access on the node does not automatically gain access to the certificates.
+
+Cons:
+* We pick up a dependency on SPIRE. We’ll need to be running a SPIRE server in the cluster somewhere, and a SPIRE agent on each node (the agent is required for the attestation process to work properly).
+
+#### Other options for SPIFFE identity
+
+##### cert-manager CSI driver
+
+cert-manager has a SPIFFE-capable CSI plugin that makes SVIDs, including the X.509 TLS keypair, available as a dynamically-mounted Secret inside Kubernetes Pods. Because we don’t want the Pods to need to mount things for mTLS, this is not viable, unfortunately.
+
+##### Istio
+
+The Istio control plane issues SPIFFE IDs to all workloads, but it seems to do it on a per-service-hostname basis, which also doesn’t match up well with the Cilium Security ID.
```

Comment:
per service account, not related to Service/hostname

Reply:
Ah, thanks, updated!